Enable optional/required features to be checked before session creation (#739)
immersive-web/webxr: explainer.md

## Inline sessions

> When authoring content to be viewed immersively, it may be beneficial to use an `inline` session to view the same content in a 2D browser window. Using an inline session enables content to use a single rendering path for both inline and immersive presentation modes. It also makes switching between inline content and immersive presentation of that content easier.

Reviewer: 👍 Thanks for re-aligning this statement with the current design priorities.

`navigator.xr.supportsSession()` will always return `true` when checking the support of `"inline"` sessions.  The UA should not reject requests for an inline session unless the page's feature policy prevents it or unless a required feature is unavailable as described in [Feature dependencies](#feature-dependencies)). For example, the following use cases all depend on additional reference space types which would need to be enabled via the `XRSessionInit`:
 - Using phone rotation to view panoramic content.
 - Taking advantage of 6DoF tracking on devices with no associated headset, like [ARCore](https://developers.google.com/ar/) or [ARKit](https://developer.apple.com/arkit/) enabled phones. (Note that this is not the same as `immersive-ar` as it does not provide automatic camera composition)
 - Making use of head-tracking features for devices like [zSpace](http://zspace.com/) systems.

## Advanced functionality

Beyond the core APIs described above, the WebXR Device API also exposes several options for taking greater advantage of the XR hardware's capabilities.

### Feature dependencies

> Once developers have mastered session creation and rendering, they will often be interested in using additional WebXR features that may not be universally available. While developers are generally encouraged to design for progressive enhancement, some experiences may have requirements on features that are not guaranteed to be universally available. For example, a guided tour would not function on an Oculus Go because it is unable to provide an [`unbounded` reference space](spatial-tracking-explainer.md#unbounded-reference-space). If an experience is completely unusable without a specific feature, it would be a poor user experience to initialize the underlying XR platform and create a session only to immediately notify the user it won't work.

Reviewer: > For example, a guided tour would not function on an Oculus Go because it is unable to provide an `unbounded` reference space.

This example sounded weird to me on first read, because a certain form of guided tour (Navigating between photospheres, street view-style) is reasonably popular on 3DoF headsets like the Go. What about: 

>  **an experience which requires users to move around a large physical space, such as** a guided tour, would not function on an Oculus Go because...

Developers communicate their feature requirements by categorizing them into one of the following groups:
* **Required** This feature must be available in order for the experience to function at all. If [explicit consent](privacy-security-explainer.md#explicit-consent) is necessary, users will be prompted in response to `xr.requestSession()`. Session creation will be rejected if the feature is unavailable for the XR device or if the UA determines the user does not wish the feature enabled.
* **Optional** The experience would like to use this feature for the entire session, but can function without it. Again, if [explicit consent](privacy-security-explainer.md#explicit-consent) is necessary, users will be prompted in response to `xr.requestSession()`. However, session creation will succeed regardless of the feature's hardware support or user intent. Developers must not assume optional features are available in the session and check the result from attempting to use them.

> Features recognized by the UA that are not explicitly listed in these arrays will be implicitly added to `XRSessionInit.optionalFeatures` if they do not require consent or [consent has been implied](privacy-security-explainer.md#implied-consent). For example, creating an `immersive-vr` session implies consent for `local` and `local-floor` reference spaces even if not explicitly listed in `XRSessionInit`. If the UA does not recognize an entry in `XRSession.requiredFeatures` the session will not be created.

Reviewer: > Features recognized by the UA that are not explicitly listed in these arrays will be implicitly added to `XRSessionInit.optionalFeatures` if they do not require consent or consent has been implied.

I'm sure how to interpret this line. Are you stating that if user agent A that has determined it can safely offer, say, eye-tracking without user consent should implicitly append it to the optional array? (*Terrible* example, I know, but work with me.) That seems problematic, because then if user agent B decides that eye-tracking really *does* require explicit consent, experiences written only against A will fail because the feature was not requested.

I'm onboard with `local` and `local-floor` being implicit for `immersive-vr`, but I feel like they probably need to be part of an explicit list of implied features rather than something left to the user agent.

Author: >I'm sure how to interpret this line. Are you stating that if user agent A that has determined it can safely offer, say, eye-tracking without user consent should implicitly append it to the optional array? (Terrible example, I know, but work with me.) That seems problematic, because then if user agent B decides that eye-tracking really does require explicit consent, experiences written only against A will fail because the feature was not requested.

Perhaps this can be fixed by changing the sentence to say "if they do not require a signal of user intent"?  For context, I added this sentence based on our noodling about the idea that maybe there are features that have no consent requirements at all, such as the ability to support an xr-standard gamepad.  At the time you were hesitant to include it because if gamepad based motion controllers are present, they should always appear in the inputSources array.  My case for including it on the list was that there are some experiences which will not work if the hardware cannot support a multibutton input source.  I tried to strike the balance by suggesting only automatically enabling features that are normatively specified to not be dependent on a signal of user intent.  Does that make more sense?

Author: > I'm onboard with local and local-floor being implicit for immersive-vr, but I feel like they probably need to be part of an explicit list of implied features rather than something left to the user agent.

My next PR calls this out a lot more concretely as it digs into threats and mitigations, but I'm ok with making some sort of a list here if it helps.  I only left it out of here because the list is already located in the [Spatial tracking explainer](spatial-tracking-explainer.md#xrreferencespace-availability)

Author: Perhaps both issues can be address with a single list of "exceptions" added to this section?

Reviewer: > Perhaps this can be fixed by changing the sentence to say "if they do not require a signal of user intent"?

Ah, I'm see what you were trying to say. I'm OK with this mental model of "everything that doesn't require user intent is automatically optional". It does help clarify the gamepad case, and I'm sure others. However, we'd still want an explicit list of which features those are so that we don't get UA disagreements.

Author: Ok, how about I switch the whole paragraph around to this:

> If the UA does not recognize an entry in `XRSession.requiredFeatures` the session will not be created. Under some circumstances, features recognized by the UA that are not explicitly listed in these arrays may be implicitly added to `XRSessionInit.optionalFeatures`. For example, if a feature does not require a signal of [user intent](privacy-security-explainer.md#user-intent). Or, if the requested session mode implies consent, such as `immersive-vr` does for `local` and `local-floor` reference spaces.

Reviewer: Better, thanks!

> If the UA does not recognize an entry in `XRSession.requiredFeatures` the session will not be created. Under some circumstances, features recognized by the UA but not explicitly listed in these arrays may be implicitly added to `XRSessionInit.optionalFeatures`. For example, if a feature does not require a signal of [user intent](privacy-security-explainer.md#user-intent). Or, if the requested session mode implies consent, such as `immersive-vr` does for `local` and `local-floor` reference spaces. 

Reviewer 2: I'd suggest adding two additional concepts here:
- Performance might affect whether a feature is added implicitly
- Which features get added implicitly might vary by user agent. Don't count on anything not `required`

Also, I think I know what the last sentence means, but the phrase 'implies consent' could be interpreted as 'implies user consent' so I'm not certain. I think the intention is that session mode could imply that certain features are available.

In case it helps, here's a suggested alternative including the above concepts and clarifying the meaning around session mode (note: this isn't intended to be blocking, I'm just including it in case it's useful):

> If the UA does not recognize an entry in `XRSession.requiredFeatures` the session will not be created.
>
> Features not listed in either array may or may not be available during the session. Under some circumstances the UA may implicitly add features to `XRSessionInit.optionalFeatures`, enabling them during the session.
>
>For example, a user agent may choose to enable a feature that does not require a signal of [user intent](privacy-security-explainer.md#user-intent) and which has no adverse performance implications. This behavior is likely to vary between user agents, and developers should not depend upon the availability of features not explicitly listed as `required`.
>
> The user agent may also enable features that are implied by session type, or implied by other requested features. For example, as any immersive session must support both `local` and `local-floor` reference spaces, a user agent may add both references spaces to `XRSessionInit.optionalFeatures` when an `immersive-vr` session is created.

On the last bit, I'm curious whether this is a firm requirement. For example, consider a scenario where a developer requests an immersive session with a `local` reference space, but not `local-floor`. Suppose the user has configured their viewer height. #638 proposes that exposing actual viewer height requires [explicit consent](privacy-security-explainer.md#explicit-consent) because it is sensitive data. In this case, should the user agent:

1. Require [explicit consent](privacy-security-explainer.md#explicit-consent) because `local-floor` is implied? 
2. Add an implicit `local-floor` reference space, but use emulated floor height to avoid [explicit consent](privacy-security-explainer.md#explicit-consent)?
3. Not add `local-floor` at all, even though it's an immersive session?
4. Any of the above, at the discretion of the user agent (and if a developer needs `local-floor` they should make it `required`?)

Author: Good point about potential performance feedback!  And I definitely didn't make it clear enough, that it is not a choice to be left to the UA discretion about which features qualify for being automatically enabled.  Without consistency on this, we'd end up in a world of hurt.  

I also, better understand your concern now about `local-floor` technically having a different threat profile than simply `local`.  While it is a requirement that `local` and `local-floor` be *possible* in any session that supports `immersive-vr` and `immersive-ar`, it doesn't mean we necessarily want to expose user height without the UA having the opportunity to manage that risk differently.

Hopefully my latest commit has addressed both your concerns and @toji's. If so, I'd appreciate an "Approve" :)

Reviewer: The table definitely helps clarify things, thanks!

To make sure I understand your above comment, Nell: Sounds like the intent is that `local-floor` is no longer one of the implicit features, but (as is currently required by the spec) would still be available to any immersive session via an emulated floor height. The only way to get a non-emulated floor height would be to explicitly request `local-floor` in either the `requiredFeatures` or `optionalFeatures` lists. Is that correct?

Author: This change came from my discussion with @johnpallett on Friday, so please correct me if I've misunderstood.  I was under the impression that we couldn't automatically allow a local-floor with even an emulated height because it could be used for fingerprinting.

Reviewer 2: The concern with `local-floor` is when the viewer height is _not_ emulated. The user's height is sensitive data (i.e. 'are you a child') so would likely require explicit consent on some platforms.

In contrast, `local-floor` with _emulated_ height doesn't have this concern, and the UA can round the data to mitigate the fingerprinting risk.

@NellWaliczek If I understood on Friday the primary concern was making sure user agents were consistent in their behavior; it seems like requiring an _emulated_ `local-floor` in this scenario could achieve that consistency.

Author: Oh, right now I remember.  We have two competing problems to solve:

1. We definitely don't want `local-floor` to be forced to use an emulated height when a real one is available (either via sensor data or a user configuration setting).
1. We definitely don't want to have inconsistency about some hardware needing to have `local-floor` requested up front and others not.  

The intent behind this approach was simply to settle on the lowest common denominator to ensure consistency and user safety.

@toji, you on board with that?

Reviewer 2: I'm not @toji but that makes sense to me and I agree with the approach as written

Reviewer: To make sure I'm on the same page: My reading of this is that `local-floor` must be explicitly requested to be used, and `local` is the only reference space that will be implicitly available to immersive sessions. If that's correct, I'm on board with it.

Author: That's correct

WebXR allows the following features to be requested:
* `local`
* `local-floor`
* `bounded-floor`
* `unbounded`

This list is currently limited to a subset of reference space types, but in the future will expand to include additional features. Some potential future features under discussion that would be candidates for this list are: eye tracking, plane detection, geo alignment, etc.

Developers communicate their feature requirements by categorizing them into one of the following groups:

> * **Required** This feature must be available in order for the experience to function at all. If [explicit consent](privacy-security-explainer.md#explicit-consent) is necessary, users will be prompted in response to `xr.requestSession()`. Session creation will be rejected if the feature is unavailable for the XR device or if the UA determines the user does not wish the feature enabled.

Reviewer 2: ```suggestion
* **Required** This feature must be available in order for the experience to function at all. If [explicit consent](privacy-security-explainer.md#explicit-consent) is necessary, users will be prompted in response to `xr.requestSession()`. Session creation will be rejected if the feature is unavailable for the XR device or if the UA determines the user does not wish the feature enabled. If the UA does not recognize an entry in `XRSession.requiredFeatures` the session will not be created. 
```

* **Optional** The experience would like to use this feature for the entire session, but can function without it. Again, if [explicit consent](privacy-security-explainer.md#explicit-consent) is necessary, users will be prompted in response to `xr.requestSession()`. However, session creation will succeed regardless of the feature's hardware support or user intent. Developers must not assume optional features are available in the session and check the result from attempting to use them.

> If the UA does not recognize an entry in `XRSession.requiredFeatures` the session will not be created. Some features recognized by the UA but not explicitly listed in these arrays may be enabled by default for a session. This is only allowed if the feature does not require a signal of [user intent](privacy-security-explainer.md#user-intent) nor impact performance or the behavior of other features when enabled. At this time, only the following features may be enabled by default:

Reviewer: > If the UA does not recognize an entry in `XRSession.requiredFeatures` the session will not be created.

Nit: Upon re-reading with fresh eyes, this sentence feels out-of-place with respect to the rest of the paragraph. Would it fit better at the end of the **Required:**  paragraph above?

Reviewer 2: @toji I agree, that sentence probably makes more sense in the 'required' section. I've added a suggestion above in case it helps.

@NellWaliczek the use of 'may' and 'allowed' here could be interpreted as user agent discretion - i.e. some UAs might add 'local' and others might not. Was that the intent? If not I'd suggest '... will be **enabled** by default...' and 'This is only **done** if the feature...' as follows:

```suggestion
Some features recognized by the UA but not explicitly listed in these arrays will be enabled by default for a session. This is only done if the feature does not require a signal of [user intent](privacy-security-explainer.md#user-intent) nor impact performance or the behavior of other features when enabled. At this time, only the following features will be enabled by default:

```

Author: >I've added a suggestion above in case it helps.

@johnpallett, above where?

Author: Oh, GitHub.  It put your comments in two places.  Found it.

  "immersive-ar"
}

> dictionary XRSessionInit {

Reviewer 3: This still needs to be added to the requestSession argument list above?

```webidl
Promise<XRSession> requestSession(XRSessionMode mode, optional XRSessionInit options);
```

Speaking of this, is the intent that these init options only apply to `requestSession`, not `supportsSession`? If yes, I think it would be helpful to call that out specifically in the explainer. For example, asking for options may require prompting the user to get an answer, and `supportsSession` is intended to be non-interactive.

Reviewer: Yes, the intent is these options only apply to `requestSession`. Not only would it violate the desire for `supportsSession` to be non-interactive to properly resolve some of the requirements, it would also make it a much more potent fingerprinting venue. As such `supportsSession` will remain `mode`-only.

Author: Thanks for catching that!
